### PR TITLE
Book security category

### PR DESCRIPTION
**Pull request by Nikname2303**

### CODE REVIEW

File: src/main/java/com/example/bookshop/dto/book/CreateBookRequestDto.java
Comment (AnyaRatusnyak) [inline src/main/java/com/example/bookshop/dto/book/CreateBookRequestDto.java:1]: Why don't you add categories in class CreateBookRequestDto? private Set<Long> categoryIds;

File: src/main/java/com/example/bookshop/mapper/BookMapper.java
Comment (AnyaRatusnyak) [inline src/main/java/com/example/bookshop/mapper/BookMapper.java:1]: I can't find method  set  categories from CreateBookRequestDto  to model book with categories

**Review by AnyaRatusnyak — APPROVED** [review]
Good job )

File: src/main/java/com/example/bookshop/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+@Tag(name = "Category management", description = "Endpoints for managing categories")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/api/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+
+    @Operation(summary = "Get all categories", description = "Get a list with all categories")
+    @GetMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/controller/CategoryController.java]:
```suggestion
    @PreAuthorize("hasRole('USER')")
```

Comment (andrii-hoienko) [reply src/main/java/com/example/bookshop/controller/CategoryController.java]: Should work without ROLE_

File: src/main/java/com/example/bookshop/dto/book/BookDtoWithoutCategoryIds.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    private String title;
+    @NotBlank
+    private String author;
+    @NotBlank
+    private String isbn;
+    @NotNull
+    @PositiveOrZero
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/dto/book/BookDtoWithoutCategoryIds.java]: What is the reason to validate it? It this class used as requestDto?

File: src/main/java/com/example/bookshop/dto/category/CategoryRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.example.bookshop.dto.category;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CategoryRequestDto {
+    private Long id;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/dto/category/CategoryRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/bookshop/dto/category/CategoryRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    @NotBlank
+    private String name;
+    private String description;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/dto/category/CategoryRequestDto.java]: Add max limit

File: src/main/java/com/example/bookshop/model/Book.java [hunk 5]
```diff
@@ -34,4 +40,11 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/model/Book.java]:
```suggestion
    @ManyToMany(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/bookshop/service/CategoryService.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+    CategoryResponseDto getById(Long id);
+
+    CategoryResponseDto save(CategoryRequestDto categoryRequestDto);
+
+    void updateById(Long id, CategoryRequestDto categoryRequestDto);
+
+    void deleteById(Long id);
+
+    public List<BookResponseDto> getBooksByCategoryId(Long id);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/CategoryService.java]:
```suggestion
    List<BootWithoutCategories> getBooksByCategoryId(Long id);
```

File: src/main/java/com/example/bookshop/service/impl/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public void updateById(Long id, CategoryRequestDto categoryRequestDto) {
+        Category category = categoryMapper.toModel(categoryRequestDto);
+        category.setId(id);
+        categoryRepository.save(category);
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        categoryRepository.deleteById(id);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/CategoryServiceImpl.java]: Please check if the entity exists before taking any actions

File: src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -19,20 +19,20 @@ public class BookServiceImpl implements BookService {
     private final BookMapper bookMapper;
 
     @Override
-    public BookDto save(CreateBookRequestDto requestDto) {
+    public BookResponseDto save(CreateBookRequestDto requestDto) {
         Book book = bookMapper.toModel(requestDto);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java:40]: You should check if book categories exist before saving

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job! Left some comments

File: src/main/java/com/example/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 9]
```diff
@@ -31,7 +31,7 @@ public ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundExcept
 
     @ExceptionHandler({Exception.class})
     public ResponseEntity<Object> handleException(Exception exception) {
-        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+        return ResponseEntity.status(HttpStatus.FORBIDDEN)
                 .body(exception.getMessage());
     }
```
Comment (yevheniisynytsia) [inline src/main/java/com/example/bookshop/exception/CustomGlobalExceptionHandler.java]: I think it is more 5xx kind of status

An Exception occurred and you don't have a method to handle it

File: src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java [hunk 10]
```diff
@@ -20,6 +20,9 @@ public class BookServiceImpl implements BookService {
 
     @Override
     public BookResponseDto save(CreateBookRequestDto requestDto) {
+        if (requestDto.getCategoryIds().isEmpty()) {
+            throw new EntityNotFoundException("Category cannot be null");
+        }
```
Comment (yevheniisynytsia) [inline src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java:29]: Andrii meant to check if the category ids that you pass in the `CreateBookRequestDto` are present in your db.
For example I may pass id 93123928381293. What will happen here and what you'd expect such an app to behave?

Let's think how to do it properly. Tip: get all the needed categories in one call to db. And make an informative exception message for the case when there is no category with such id(s)

File: src/main/java/com/example/bookshop/model/Book.java [hunk 11]
```diff
@@ -34,4 +40,11 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (yevheniisynytsia) [inline src/main/java/com/example/bookshop/model/Book.java:43]: are you able to get a book with its categories properly with `LAZY`? Have you tested it?

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java [hunk 12]
```diff
@@ -16,13 +18,25 @@
...
         if (requestDto.getCategoryIds().isEmpty()) {
             throw new EntityNotFoundException("Category cannot be null");
         }
+        List<Category> categories = categoryRepository.findAllById(requestDto.getCategoryIds());
+        List<Long> foundCategoryIds = categories.stream()
+                .map(Category::getId)
+                .toList();
+        List<Long> missingCategoryIds = requestDto.getCategoryIds().stream()
+                .filter(id -> !foundCategoryIds.contains(id))
+                .toList();
```
Comment (yevheniisynytsia) [inline src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java]: I think `removeAll` instead of `iterating+contains `may work faster here

**Review by yevheniisynytsia — APPROVED** [review]

**Pull request merged**